--- wbxml2-0.9.2/src/wbxml_tree.h.orig	2008-10-22 20:28:03.167549767 +0200
+++ wbxml2-0.9.2/src/wbxml_tree.h	2008-10-22 20:29:27.661742674 +0200
@@ -194,7 +194,8 @@
                                             WBXMLGenXMLParams *params);
 
 /** @todo Libxml support ! */
-#if defined( HAVE_LIBXML )
+//#if defined( HAVE_LIBXML )
+#if 0
 
 /**
  * @brief Parse a LibXML document, and construct the corresponding WBXML Tree
--- wbxml2-0.9.2/libwbxml2.pc.in.orig	2008-10-22 20:30:16.658375247 +0200
+++ wbxml2-0.9.2/libwbxml2.pc.in	2008-10-22 20:32:22.149750512 +0200
@@ -6,6 +6,5 @@
 Name: libwbxml2
 Description: C wbxml library
 Version: @VERSION@
-Requires: libxml-2.0 >= 2.6
 Libs: -L${libdir} -lwbxml2
 Cflags: -I${includedir} -DHAVE_EXPAT -D_REENTRANT -DWBXML_ENCODER_USE_STRTBL -DWBXML_SUPPORT_WML -DWBXML_SUPPORT_WTA -DWBXML_SUPPORT_SI -DWBXML_SUPPORT_SL -DWBXML_SUPPORT_CO -DWBXML_SUPPORT_PROV -DWBXML_SUPPORT_EMN -DWBXML_SUPPORT_DRMREL -DWBXML_SUPPORT_OTA_SETTINGS -DWBXML_SUPPORT_SYNCML -DWBXML_SUPPORT_WV -DWBXML_SUPPORT_AIRSYNC
