Index: /wbxml2/trunk/src/wbxml_tables.h
===================================================================
--- /wbxml2/trunk/src/wbxml_tables.h (revision 27)
+++ /wbxml2/trunk/src/wbxml_tables.h (revision 41)
@@ -109,6 +109,6 @@
 /* SyncML 1.2 */
 #define WBXML_PUBLIC_ID_SYNCML_SYNCML12 0x1201
-#define WBXML_PUBLIC_ID_SYNCML_DEVINF12 0x1202
-#define WBXML_PUBLIC_ID_SYNCML_METINF12 0x1203
+#define WBXML_PUBLIC_ID_SYNCML_METINF12 0x1202
+#define WBXML_PUBLIC_ID_SYNCML_DEVINF12 0x1203
 
 #define XML_PUBLIC_ID_SYNCML_SYNCML12 "-//SYNCML//DTD SyncML 1.2//EN"
Index: /wbxml2/trunk/src/wbxml_encoder.c
===================================================================
--- /wbxml2/trunk/src/wbxml_encoder.c (revision 39)
+++ /wbxml2/trunk/src/wbxml_encoder.c (revision 41)
@@ -1937,5 +1937,6 @@
         /* If this is a SyncML document ? */
         if ((encoder->lang->langID == WBXML_LANG_SYNCML_SYNCML10) ||
-            (encoder->lang->langID == WBXML_LANG_SYNCML_SYNCML11))
+            (encoder->lang->langID == WBXML_LANG_SYNCML_SYNCML11) ||
+            (encoder->lang->langID == WBXML_LANG_SYNCML_SYNCML12))
         {
             /** @todo We must check too if we are in a <Type> */
@@ -4020,5 +4021,6 @@
         /* Change text in <Type> from "application/vnd.syncml-devinf+wbxml" to "application/vnd.syncml-devinf+xml" */
         if (((encoder->lang->langID == WBXML_LANG_SYNCML_SYNCML10) ||
-             (encoder->lang->langID == WBXML_LANG_SYNCML_SYNCML11)) &&
+             (encoder->lang->langID == WBXML_LANG_SYNCML_SYNCML11) ||
+             (encoder->lang->langID == WBXML_LANG_SYNCML_SYNCML12)) &&
             (encoder->current_tag != NULL) &&
             (encoder->current_tag->wbxmlCodePage == 0x01 ) &&
