Index: /wbxml2/trunk/configure.in
===================================================================
--- /wbxml2/trunk/configure.in (revision 52)
+++ /wbxml2/trunk/configure.in (revision 58)
@@ -1,5 +1,5 @@
 dnl  Process this file with autoconf to produce a configure script.
-AC_INIT(doxygen.h)
-AM_INIT_AUTOMAKE(wbxml2, 0.9.0)
+AC_INIT(src/wbxml.h)
+AM_INIT_AUTOMAKE(wbxml2, 0.9.2)
 AM_CONFIG_HEADER(config.h)
 
Index: /wbxml2/trunk/ChangeLog
===================================================================
--- /wbxml2/trunk/ChangeLog (revision 57)
+++ /wbxml2/trunk/ChangeLog (revision 58)
@@ -1,5 +1,7 @@
 2008-02-28  Aymerick Jehanne <aymerick@jehanne.org>
-  * Applied wbxml-svn-r53-anonymous.patch from synce project:
-    https://synce.svn.sourceforge.net/svnroot/synce/trunk/patches/
+  * Applied wbxml-svn-r53-anonymous.patch from synce project.
+  * Applied wbxml-svn-r53-build.patch from synce project. Blindly, because I have no
+    linux box anymore to test. So if you have issues when compiling on linux, send me
+    a mail please.
 
 2006-07-10  Aymerick Jehanne <aymerick@jehanne.org>
Index: /wbxml2/trunk/src/Makefile.am
===================================================================
--- /wbxml2/trunk/src/Makefile.am (revision 27)
+++ /wbxml2/trunk/src/Makefile.am (revision 58)
@@ -4,5 +4,5 @@
 	 -I. -I/usr/include
 
-CFLAGS =\
+AM_CFLAGS =\
 	 -DHAVE_CONFIG_H\
 	 -DHAVE_EXPAT\
@@ -21,7 +21,5 @@
 	 -DWBXML_SUPPORT_WV\
 	 -DWBXML_SUPPORT_AIRSYNC\
-	 -Wall -Wimplicit -Wreturn-type -Wunused -Wswitch -Wcomment -Wuninitialized -Wparentheses -Wpointer-arith -Wmissing-prototypes\
-	 -O3\
-	 -g
+	 -Wall
 
 lib_LTLIBRARIES = libwbxml2.la
Index: /wbxml2/trunk/tools/Makefile.am
===================================================================
--- /wbxml2/trunk/tools/Makefile.am (revision 27)
+++ /wbxml2/trunk/tools/Makefile.am (revision 58)
@@ -4,5 +4,5 @@
 	 -I. -I../src -I/usr/include
 
-CFLAGS =\
+AM_CFLAGS =\
 	 -DHAVE_CONFIG_H\
 	 -DHAVE_EXPAT\
@@ -21,7 +21,5 @@
 	 -DWBXML_SUPPORT_WV\
 	 -DWBXML_SUPPORT_AIRSYNC\
-	 -Wall -Wimplicit -Wreturn-type -Wunused -Wswitch -Wcomment -Wuninitialized -Wparentheses -Wpointer-arith -Wmissing-prototypes\
-	 -O3\
-	 -g
+	 -Wall
 
 bin_PROGRAMS = wbxml2xml xml2wbxml
Index: /wbxml2/trunk/libwbxml2.pc.in
===================================================================
--- /wbxml2/trunk/libwbxml2.pc.in (revision 14)
+++ /wbxml2/trunk/libwbxml2.pc.in (revision 58)
@@ -9,3 +9,3 @@
 Requires: libxml-2.0 >= 2.6
 Libs: -L${libdir} -lwbxml2
-Cflags: -I${includedir}
+Cflags: -I${includedir} -DHAVE_EXPAT -D_REENTRANT -DWBXML_ENCODER_USE_STRTBL -DWBXML_SUPPORT_WML -DWBXML_SUPPORT_WTA -DWBXML_SUPPORT_SI -DWBXML_SUPPORT_SL -DWBXML_SUPPORT_CO -DWBXML_SUPPORT_PROV -DWBXML_SUPPORT_EMN -DWBXML_SUPPORT_DRMREL -DWBXML_SUPPORT_OTA_SETTINGS -DWBXML_SUPPORT_SYNCML -DWBXML_SUPPORT_WV -DWBXML_SUPPORT_AIRSYNC
Index: /wbxml2/trunk/Makefile.am
===================================================================
--- /wbxml2/trunk/Makefile.am (revision 53)
+++ /wbxml2/trunk/Makefile.am (revision 58)
@@ -22,6 +22,5 @@
 	References\
 	THANKS\
-	TODO\
-	doxygen.h
+	TODO
 
 dist-bz2: distdir
Index: /wbxml2/trunk/bootstrap
===================================================================
--- /wbxml2/trunk/bootstrap (revision 5)
+++ /wbxml2/trunk/bootstrap (revision 58)
@@ -1,8 +1,2 @@
-#! /bin/sh
-libtoolize
-aclocal
-autoheader
-#  automake --foreign --add-missing --copy
-automake --add-missing
-autoconf
-./configure --prefix=/usr
+#!/bin/sh
+autoreconf -i
