Index: /wbxml2/trunk/src/wbxml_encoder.c
===================================================================
--- /wbxml2/trunk/src/wbxml_encoder.c (revision 41)
+++ /wbxml2/trunk/src/wbxml_encoder.c (revision 42)
@@ -1500,6 +1500,6 @@
             else {
 #endif /* WBXML_ENCODER_USE_STRTBL */
-                /* Length of String Table is length of XML Public ID */
-                strstbl_len = wbxml_buffer_len(pid);
+                /* Length of String Table is length of XML Public ID (plus the NULL char) */
+                strstbl_len = wbxml_buffer_len(pid) + 1;
 
                 /* There is only the XML Public ID in String Table */
@@ -1546,8 +1546,10 @@
 
         if (pid != NULL) {
-            /* The append includes terminating NULL char */
             if (!wbxml_buffer_append(header, pid))
                 return WBXML_ERROR_ENCODER_APPEND_DATA;
             
+            if (!wbxml_buffer_append_char(header, WBXML_STR_END))
+                return WBXML_ERROR_ENCODER_APPEND_DATA;
+
             /* Clean up */
             wbxml_buffer_destroy(pid);
