Index: /wbxml2/trunk/src/wbxml_encoder.c
===================================================================
--- /wbxml2/trunk/src/wbxml_encoder.c (revision 13)
+++ /wbxml2/trunk/src/wbxml_encoder.c (revision 34)
@@ -638,4 +638,8 @@
 WBXML_DECLARE(WBXMLError) wbxml_encoder_encode_node(WBXMLEncoder *encoder, WBXMLTreeNode *node)
 {
+    if (encoder->flow_mode == FALSE) {
+        WBXML_WARNING((WBXML_ENCODER, "You should NOT call wbxml_encoder_encode_node() if you are not in Flow Mode encoding ! (use wbxml_encoder_set_flow_mode(encoder, TRUE))"));
+    }
+
     return wbxml_encoder_encode_node_with_elt_end(encoder, node, TRUE);
 }
@@ -649,8 +653,4 @@
     if ((encoder == NULL) || (node == NULL))
         return WBXML_ERROR_BAD_PARAMETER;
-    
-    if (encoder->flow_mode == FALSE) {
-        WBXML_WARNING((WBXML_ENCODER, "You should NOT call wbxml_encoder_encode_node() if you are not in Flow Mode encoding ! (use wbxml_encoder_set_flow_mode(encoder, TRUE))"));
-    }
     
     /* Check that language table has been set */
