Index: /wbxml2/trunk/src/wbxml_tree.c
===================================================================
--- /wbxml2/trunk/src/wbxml_tree.c (revision 35)
+++ /wbxml2/trunk/src/wbxml_tree.c (revision 49)
@@ -257,5 +257,6 @@
 #else /* HAVE_EXPAT */
 
-#if defined( HAVE_LIBXML )
+//#if defined( HAVE_LIBXML )
+#if 0
 
     /** @todo Use LibXML2 SAX interface ! */
@@ -334,5 +335,6 @@
 
 
-#if defined( HAVE_LIBXML )
+// #if defined( HAVE_LIBXML )
+#if 0
 
 WBXML_DECLARE(WBXMLError) wbxml_tree_from_libxml_doc(xmlDocPtr libxml_doc,
