Index: /wbxml2/trunk/src/wbxml_parser.c
===================================================================
--- /wbxml2/trunk/src/wbxml_parser.c (revision 15)
+++ /wbxml2/trunk/src/wbxml_parser.c (revision 43)
@@ -660,10 +660,12 @@
      * (for example, if the version is 0x02, then parser->version will be 0xffffff02)
      */
-    parser->version = WBXML_VERSION_10;
-    
-    if ((ret = parse_uint8(parser, (WB_UTINY*) &parser->version)) != WBXML_OK)
+    WB_UTINY version = WBXML_VERSION_10;
+    
+    if ((ret = parse_uint8(parser, &version)) != WBXML_OK)
         return ret;
-
-    WBXML_DEBUG((WBXML_PARSER, "(%d) Parsed version: '0x%X'", parser->pos - 1, (WB_TINY) parser->version));
+    
+    parser->version = version;
+
+    WBXML_DEBUG((WBXML_PARSER, "(%d) Parsed version: 1.%d", parser->pos - 1, parser->version));
 
     return WBXML_OK;
@@ -939,4 +941,6 @@
         return ret;
     }
+
+    WBXML_DEBUG((WBXML_PARSER, "<%s>", wbxml_tag_get_xml_name(element)));
   
     /* Set Current Tag */
@@ -991,5 +995,5 @@
                                                is_empty);
     }
-      
+    
     /* Free Attributes */
     free_attrs_table(attrs);
@@ -1024,5 +1028,5 @@
         }
         
-        WBXML_DEBUG((WBXML_PARSER, "(%d) End of Element", parser->pos));
+        WBXML_DEBUG((WBXML_PARSER, "(%d) End of Element", parser->pos - 1));
         
         /* Skip END */
@@ -1038,5 +1042,7 @@
                                              is_empty);
     }
-      
+    
+    WBXML_DEBUG((WBXML_PARSER, "</%s>", wbxml_tag_get_xml_name(element)));
+    
     /* Free Tag */
     wbxml_tag_destroy(element);
@@ -1079,5 +1085,5 @@
 
     if ((WB_UTINY) parser->version < (WB_UTINY) WBXML_VERSION_12)
-        WBXML_WARNING((WBXML_PARSER, "No Switch Page mecanism possible in WBXML < %s", WBXML_VERSION_TEXT_12));
+        WBXML_WARNING((WBXML_PARSER, "No Switch Page mecanism possible in WBXML < %s (current: %d)", WBXML_VERSION_TEXT_12, (WB_UTINY) parser->version));
 
     /* Skip SWITCH_PAGE token */
