Index: /wbxml2/trunk/test/test_parser.c
===================================================================
--- /wbxml2/trunk/test/test_parser.c (revision 5)
+++ /wbxml2/trunk/test/test_parser.c (revision 44)
@@ -40,5 +40,5 @@
 
 /** Start Document Callback */
-void parse_clb_start_document(void *ctx, WB_LONG charset, const WBXMLLangEntry *lang)
+void parse_clb_start_document(void *ctx, WBXMLCharsetMIBEnum charset, const WBXMLLangEntry *lang)
 {
     printf("Parsing Document:\n"
Index: /wbxml2/trunk/src/wbxml_parser.c
===================================================================
--- /wbxml2/trunk/src/wbxml_parser.c (revision 43)
+++ /wbxml2/trunk/src/wbxml_parser.c (revision 44)
@@ -278,5 +278,5 @@
       
             WBXML_WARNING((WBXML_PARSER,
-                           "No charset information found, using default : %s",
+                           "No charset information found, using default : %x",
                            WBXML_PARSER_DEFAULT_CHARSET));
         }
@@ -999,5 +999,5 @@
     free_attrs_table(attrs);
     
-      
+    
     /* Parse *content */
     if (!is_empty) {
